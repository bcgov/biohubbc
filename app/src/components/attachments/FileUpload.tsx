import Box from '@material-ui/core/Box';
import List from '@material-ui/core/List';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import makeStyles from '@material-ui/core/styles/makeStyles';
import React, { useEffect, useState } from 'react';
import { FileError, FileRejection } from 'react-dropzone';
import DropZone, { IDropZoneConfigProps } from './DropZone';
import {
  IFileHandler,
  IOnUploadSuccess,
  IUploadHandler,
  MemoizedFileUploadItem,
  UploadFileStatus
} from './FileUploadItem';

const useStyles = makeStyles((theme: Theme) => ({
  dropZone: {
    clear: 'both',
    borderRadius: '4px',
    borderStyle: 'dashed',
    borderWidth: '2px',
    borderColor: theme.palette.text.disabled,
    background: theme.palette.primary.main + '11',
    transition: 'all ease-out 0.2s',
    '&:hover, &:focus': {
      borderColor: theme.palette.primary.main,
      backgroundColor: theme.palette.primary.main + '22'
    },
    cursor: 'pointer'
  }
}));

export interface IUploadFile {
  file: File;
  error?: string;
}

export interface IUploadFileListProps {
  files: IUploadFile[];
}

export type IReplaceHandler = () => void;

export interface IFileUploadProps {
  /**
   * Callback fired for each file in the list
   *
   * @type {IUploadHandler}
   * @memberof IFileUploadProps
   */
  uploadHandler: IUploadHandler;
  /**
   * Callback fired for each accepted file in the list (that do not have any `DropZone` errors (size, count, extension).
   *
   * @type {IFileHandler}
   * @memberof IFileUploadProps
   */
  fileHandler?: IFileHandler;
  /**
   * Callback fired when `uploadHandler` runs successfully fora  given file. Will run once for each file that is
   * uploaded.
   *
   * @type {IOnUploadSuccess}
   * @memberof IFileUploadProps
   */
  onSuccess?: IOnUploadSuccess;
  /**
   * Manually dictate the status.
   *
   * Note: some component events are automatically triggered based on a change of status.
   *
   * @type {UploadFileStatus}
   * @memberof IFileUploadProps
   */
  status?: UploadFileStatus;
  /**
   * If the component should replace the selected files, rather than appending them.
   * Default: false
   *
   * Example:
   * - WIth replace=false, selecting FileA and then selecting FileB will result in both FileA and FileB in the upload
   * list.
   * - With replace=true, selecting FileA and then selecting FileB will result in only FileB in the upload list.
   *
   * Note: This will not change how many files are uploaded, only how many files appear in the list. So if
   * a file is in the middle of uploading when it is replaced, that file will still continue to upload even though it
   * is not visible in the upload list.
   *
   * @type {boolean}
   * @memberof IFileUploadProps
   */
  replace?: boolean;
  /**
   * Callback fired when files are replaced.
   *
   * Note: Does nothing if `replace` is not set to `true`.
   *
   * @type {IReplaceHandler}
   * @memberof IFileUploadProps
   */
  onReplace?: IReplaceHandler;
  dropZoneProps?: Partial<IDropZoneConfigProps>;
}

export const FileUpload: React.FC<IFileUploadProps> = (props) => {
  const classes = useStyles();

  const [files, setFiles] = useState<IUploadFile[]>([]);

  const [fileUploadItems, setFileUploadItems] = useState<any[]>([]);

  const [fileToRemove, setFileToRemove] = useState<string>('');

  /**
   * Handles files which are added (via either drag/drop or browsing).
   *
   * @param {File[]} filesToAdd files that pass the basic DropZone size/quantity/type checks
   * @param {FileRejection[]} rejectedFiles files that did not pass the basic DropZone size/quantity/type checks
   */
  const onFiles = (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
    const newAcceptedFiles: IUploadFile[] = [];
    const newRejectedFiles: IUploadFile[] = [];

    // Parse out any files that have already been added
    acceptedFiles.forEach((item) => {
      const isAlreadyAdded = files.some((existingFile) => existingFile.file.name === item.name);

      if (isAlreadyAdded) {
        return;
      }

      newAcceptedFiles.push({
        file: item
      });
    });

    // Parse out any rejected files that have already been added
    rejectedFiles.forEach((item) => {
      const isAlreadyRejected = files.some((existingFile) => existingFile.file.name === item.file.name);

      if (isAlreadyRejected) {
        return;
      }

      newRejectedFiles.push({
        file: item.file,
        error: getErrorCodeMessage(item.errors[0])
      });
    });

    if (props.replace) {
      // Replace current files with new files
      setFiles([...newAcceptedFiles, ...newRejectedFiles]);
      setFileUploadItems([
        ...newAcceptedFiles.map((item) => getFileUploadItem(item.file, item.error)),
        ...newRejectedFiles.map((item) => getFileUploadItem(item.file, item.error))
      ]);
      props.onReplace?.();
    } else {
      // Append new files to current files
      setFiles((currentFiles) => [...currentFiles, ...newAcceptedFiles, ...newRejectedFiles]);
      setFileUploadItems(
        fileUploadItems.concat([
          ...newAcceptedFiles.map((item) => getFileUploadItem(item.file, item.error)),
          ...newRejectedFiles.map((item) => getFileUploadItem(item.file, item.error))
        ])
      );
    }
  };

  const getFileUploadItem = (file: File, error?: string) => {
    return (
      <MemoizedFileUploadItem
        key={file.name}
        uploadHandler={props.uploadHandler}
        onSuccess={props.onSuccess}
        file={file}
        error={error}
        onCancel={() => setFileToRemove(file.name)}
        fileHandler={props.fileHandler}
        status={props.status}
      />
    );
  };

  const getErrorCodeMessage = (fileError: FileError) => {
    switch (fileError.code) {
      case 'file-too-large':
        return 'File size exceeds maximum';
      case 'too-many-files':
        return 'Number of files uploaded at once exceeds maximum';
      case 'file-invalid-type':
        return 'File type is not compatible';
      default:
        return 'Encountered an unexpected error';
    }
  };

  useEffect(() => {
    if (!fileToRemove) {
      return;
    }

    const removeFile = (fileName: string) => {
      const index = files.findIndex((item) => item.file.name === fileName);

      if (index === -1) {
        return;
      }

      setFiles((currentFiles) => {
        const newFiles = [...currentFiles];
        newFiles.splice(index, 1);
        return newFiles;
      });

      setFileUploadItems((currentFileUploadItems) => {
        const newFileUploadItems = [...currentFileUploadItems];
        newFileUploadItems.splice(index, 1);
        return newFileUploadItems;
      });

      setFileToRemove('');
    };

    removeFile(fileToRemove);
  }, [fileToRemove, fileUploadItems, files]);

  return (
    <Box>
      <Box className={classes.dropZone}>
        <DropZone onFiles={onFiles} {...props.dropZoneProps} />
      </Box>
      <Box>
        <List>{fileUploadItems}</List>
      </Box>
    </Box>
  );
};

export default FileUpload;
