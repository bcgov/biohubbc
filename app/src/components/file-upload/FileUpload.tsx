import Box from '@mui/material/Box';
import { grey } from '@mui/material/colors';
import List from '@mui/material/List';
import React, { useRef, useState } from 'react';
import { FileError, FileRejection } from 'react-dropzone';
import DropZone, { IDropZoneConfigProps } from './DropZone';
import FileUploadItem, {
  IFileHandler,
  IFileUploadItemProps,
  IOnUploadSuccess,
  IUploadHandler,
  UploadFileStatus
} from './FileUploadItem';
import FileUploadItemActionButton from './FileUploadItemActionButton';
import FileUploadItemProgressBar from './FileUploadItemProgressBar';

export interface IUploadFile {
  file: File;
  error?: string;
}

export type IReplaceHandler = () => void;

export interface IFileUploadProps {
  /**
   * Callback fired for each file in the list
   *
   * @type {IUploadHandler}
   * @memberof IFileUploadProps
   */
  uploadHandler?: IUploadHandler;
  /**
   * Callback fired for each accepted file in the list (that do not have any `DropZone` errors (size, count, extension).
   *
   * @type {IFileHandler}
   * @memberof IFileUploadProps
   */
  fileHandler?: IFileHandler;
  /**
   * Callback fired when `uploadHandler` runs successfully for a given file. Will run once for each file that is
   * uploaded.
   *
   * @type {IOnUploadSuccess}
   * @memberof IFileUploadProps
   */
  onSuccess?: IOnUploadSuccess;
  /**
   * Manually dictate the status.
   *
   * Note: some component events are automatically triggered based on a change of status.
   *
   * @type {UploadFileStatus}
   * @memberof IFileUploadProps
   */
  status?: UploadFileStatus;
  /**
   * If the component should replace the selected files, rather than appending them.
   * Default: false
   *
   * Example:
   * - WIth replace=false, selecting FileA and then selecting FileB will result in both FileA and FileB in the upload
   * list.
   * - With replace=true, selecting FileA and then selecting FileB will result in only FileB in the upload list.
   *
   * Note: This will not change how many files are uploaded, only how many files appear in the list. So if
   * a file is in the middle of uploading when it is replaced, that file will still continue to upload even though it
   * is not visible in the upload list.
   *
   * @type {boolean}
   * @memberof IFileUploadProps
   */
  replace?: boolean;
  /**
   * Callback fired when files are replaced.
   *
   * Note: Does nothing if `replace` is not set to `true`.
   *
   * @type {IReplaceHandler}
   * @memberof IFileUploadProps
   */
  onReplace?: IReplaceHandler;
  /**
   * Callback fired when files are removed.
   *
   * @memberof IFileUploadProps
   */
  onRemove?: (fileName: string) => void;
  /**
   * If `true`, hides the drop zone when the maximum number of allowed files has been reached.
   *
   * @type {boolean}
   * @memberof IFileUploadProps
   */
  hideDropZoneOnMaxFiles?: boolean;
  /**
   * Optional drop zone props.
   *
   * @type {Partial<IDropZoneConfigProps>}
   * @memberof IFileUploadProps
   */
  dropZoneProps?: Partial<IDropZoneConfigProps>;
  /**
   * If `true`, show advanced error details on a failed upload, for each upload item.
   *
   * @type {boolean}
   * @memberof IFileUploadProps
   */
  enableErrorDetails?: boolean;
  /**
   * A component that renders a file Upload item.
   *
   * @memberof IFileUploadProps
   */
  FileUploadItemComponent?: (props: IFileUploadItemProps) => JSX.Element;
  /**
   * Optional prop overrides for `FileUploadItemComponent`.
   *
   * @type {Partial<IFileUploadItemProps>}
   * @memberof IFileUploadProps
   */
  FileUploadItemComponentProps?: Partial<IFileUploadItemProps>;
}

export const FileUpload = (props: IFileUploadProps) => {
  const files = useRef<IUploadFile[]>([]);

  const [fileUploadItems, setFileUploadItems] = useState<any[]>([]);

  const MemoizedFileUploadItem = React.memo(props.FileUploadItemComponent ?? FileUploadItem, (prevProps, nextProps) => {
    return prevProps.file.name === nextProps.file.name;
  });

  /**
   * Handles files which are added (via either drag/drop or browsing).
   *
   * @param {File[]} filesToAdd files that pass the basic DropZone size/quantity/type checks
   * @param {FileRejection[]} rejectedFiles files that did not pass the basic DropZone size/quantity/type checks
   */
  const onFiles = (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
    const newAcceptedFiles: IUploadFile[] = [];
    const newRejectedFiles: IUploadFile[] = [];

    // Parse out any files that have already been added
    acceptedFiles.forEach((item) => {
      const isAlreadyAdded = files.current.some((existingFile) => existingFile.file.name === item.name);

      if (isAlreadyAdded) {
        return;
      }

      newAcceptedFiles.push({
        file: item
      });
    });

    // Parse out any rejected files that have already been added
    rejectedFiles.forEach((item) => {
      const isAlreadyRejected = files.current.some((existingFile) => existingFile.file.name === item.file.name);

      if (isAlreadyRejected) {
        return;
      }

      newRejectedFiles.push({
        file: item.file,
        error: getErrorCodeMessage(item.errors[0])
      });
    });

    if (props.replace) {
      // Replace current files with new files
      files.current = [...newAcceptedFiles, ...newRejectedFiles];
      setFileUploadItems([
        ...newAcceptedFiles.map((item) => getFileUploadItem(item.file, item.error)),
        ...newRejectedFiles.map((item) => getFileUploadItem(item.file, item.error))
      ]);
      props.onReplace?.();
    } else {
      // Append new files to current files
      files.current = [...files.current, ...newAcceptedFiles, ...newRejectedFiles];
      setFileUploadItems(
        fileUploadItems.concat([
          ...newAcceptedFiles.map((item) => getFileUploadItem(item.file, item.error)),
          ...newRejectedFiles.map((item) => getFileUploadItem(item.file, item.error))
        ])
      );
    }
  };

  const getFileUploadItem = (file: File, error?: string) => {
    return (
      <MemoizedFileUploadItem
        key={file.name}
        uploadHandler={props.uploadHandler}
        onSuccess={props.onSuccess}
        file={file}
        error={error}
        onCancel={() => removeFile(file.name)}
        fileHandler={props.fileHandler}
        status={props.status}
        enableErrorDetails={props.enableErrorDetails}
        ActionButtonComponent={FileUploadItemActionButton}
        ProgressBarComponent={FileUploadItemProgressBar}
        {...props.FileUploadItemComponentProps}
      />
    );
  };

  const getErrorCodeMessage = (fileError: FileError) => {
    switch (fileError.code) {
      case 'file-too-large':
        return 'File size exceeds maximum';
      case 'too-many-files':
        return 'Number of files uploaded at once exceeds maximum';
      case 'file-invalid-type':
        return 'File type is not compatible';
      default:
        return 'Encountered an unexpected error';
    }
  };

  const removeFile = (fileName: string) => {
    // Find index of file to remove
    const index = files.current.findIndex((item) => item.file.name === fileName);

    if (index === -1) {
      return;
    }

    // Update array of files
    const newFiles = [...files.current];
    newFiles.splice(index, 1);
    files.current = newFiles;

    // Update array of file item components
    setFileUploadItems((currentFileUploadItems) => {
      const newFileUploadItems = [...currentFileUploadItems];
      newFileUploadItems.splice(index, 1);
      return newFileUploadItems;
    });

    // If provided, call parent onRemove callback
    props.onRemove?.(fileName);
  };

  // Whether or not to hide the drop zone when the maximum number of allowed files has been reached
  const hideDropZoneOnMaxFiles =
    props.hideDropZoneOnMaxFiles &&
    props.dropZoneProps?.maxNumFiles &&
    files.current.length >= props.dropZoneProps.maxNumFiles;

  return (
    <Box width={'100%'}>
      {!hideDropZoneOnMaxFiles && (
        <Box
          sx={{
            clear: 'both',
            borderRadius: '6px',
            borderStyle: 'dashed',
            borderWidth: '2px',
            borderColor: grey[500],
            transition: 'all ease-out 0.2s',
            '&:hover, &:focus': {
              borderColor: 'primary.main',
              backgroundColor: grey[100]
            },
            cursor: 'pointer'
          }}>
          <DropZone onFiles={onFiles} {...props.dropZoneProps} />
        </Box>
      )}
      <Box>
        <List>{fileUploadItems}</List>
      </Box>
    </Box>
  );
};

export default FileUpload;
