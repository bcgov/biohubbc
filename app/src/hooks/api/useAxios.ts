import { useKeycloak } from '@react-keycloak/web';
import axios, { AxiosError, AxiosInstance, AxiosResponse } from 'axios';
import { useMemo } from 'react';
import { ensureProtocol } from 'utils/Utils';

export class APIError extends Error {
  status: number;
  errors?: (string | object)[];
  requestURL?: string;

  constructor(error: AxiosError) {
    super(error.response?.data?.message || error.message);

    this.name = error.response?.data?.name || error.name;
    this.status = error.response?.data?.status || error.response?.status;
    this.errors = error.response?.data?.errors || [];

    this.requestURL = `${error?.config?.baseURL}${error?.config?.url}`;
  }
}

/**
 * Returns an instance of axios with baseURL and authorization headers set.
 *
 * @return {*} {AxiosInstance} an instance of axios
 */
const useAxios = (baseUrl?: string): AxiosInstance => {
  const { keycloak } = useKeycloak();

  return useMemo(() => {
    const instance = axios.create({
      headers: {
        Authorization: `Bearer ${keycloak?.token}`
      },
      baseURL: baseUrl && ensureProtocol(baseUrl)
    });

    instance.interceptors.response.use(
      (response: AxiosResponse) => {
        return response;
      },
      (error: AxiosError) => {
        return Promise.reject(new APIError(error));
      }
    );

    return instance;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [keycloak, keycloak?.token]);
};

export default useAxios;
