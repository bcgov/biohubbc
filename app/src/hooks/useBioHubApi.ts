import { useKeycloak } from '@react-keycloak/web';
import axios from 'axios';
import { IProject, IProjectPostObject } from 'interfaces/project-interfaces';
import { IActivity, ICreateActivity, ICreateProjectResponse, ITemplate } from 'interfaces/useBioHubApi-interfaces';
import moment from 'moment';
import { useMemo } from 'react';

const API_HOST = process.env.REACT_APP_API_HOST;
const API_PORT = process.env.REACT_APP_API_PORT;

const API_URL = (API_PORT && `${API_HOST}:${API_PORT}`) || API_HOST || 'api-dev-biohubbc.apps.silver.devops.gov.bc.ca';

/**
 * Returns an instance of axios with baseURL and authorization headers set.
 *
 * @return {*}
 */
const useApi = () => {
  const { keycloak } = useKeycloak();
  const instance = useMemo(() => {
    return axios.create({
      headers: {
        Authorization: `Bearer ${keycloak?.token}`
      },
      baseURL: API_URL
    });
  }, [keycloak]);

  return instance;
};

/**
 * Returns a set of supported api methods.
 *
 * @return {object} object whose properties are supported api methods.
 */
export const useBiohubApi = () => {
  const api = useApi();

  /**
   * Get all projects.
   *
   * @return {*}  {Promise<IProject[]>}
   */
  const getProjects = async (): Promise<IProject[]> => {
    // const { data } = await api.get(`/api/projects`);

    // return data;

    // TODO: stub for development
    return [
      {
        id: 1,
        name: 'Project Name 1',
        objectives: 'Project Objectives 1',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 1',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 1',
        caveats: 'Caveats 1',
        comments: 'Comments 1'
      },
      {
        id: 2,
        name: 'Project Name 2',
        objectives: 'Project Objectives 2',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 2',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 2',
        caveats: 'Caveats 2',
        comments: 'Comments 2'
      },
      {
        id: 3,
        name: 'Project Name 3',
        objectives: 'Project Objectives 3',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 3',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 3',
        caveats: 'Caveats 3',
        comments: 'Comments 3'
      },
      {
        id: 4,
        name: 'Project Name 4',
        objectives: 'Project Objectives 4',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 4',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 4',
        caveats: 'Caveats 4',
        comments: 'Comments 4'
      }
    ];
  };

  /**
   * Get a project based on its ID.
   *
   * @param {projectId} projectId
   * @return {*}  {Promise<IProject>}
   */
  const getProject = async (projectId: number): Promise<IProject> => {
    const { data } = await api.get(`/api/project/${projectId}`);

    return data;
  };

  /**
   * Create a new project.
   *
   * @param {IProjectPostObject} project
   * @return {*}  {Promise<ICreateProjectResponse>}
   */
  const createProject = async (project: IProjectPostObject): Promise<ICreateProjectResponse> => {
    const { data } = await api.post('/api/project', project);

    return data;
  };

  /**
   * Get a template based on its ID.
   *
   * @param {templateId} templateId
   * @return {*}  {Promise<ITemplate>}
   */
  const getTemplate = async (templateId: number): Promise<ITemplate> => {
    const { data } = await api.get(`/api/template/${templateId}`);

    return data;
  };

  /**
   * Create a new activity record.
   *
   * @param {ICreateActivity} activity
   * @return {*}  {Promise<IActivity>}
   */
  const createActivity = async (activity: ICreateActivity): Promise<IActivity> => {
    const { data } = await api.post('/api/activity', activity);

    return data;
  };

  /**
   * Fetch all code sets.
   *
   * @return {*}  {Promise<any>}
   */
  const getAllCodes = async (): Promise<any> => {
    const { data } = await api.get('/api/codes/');

    return data;
  };

  /**
   * Fetch the api json-schema spec.
   *
   * @return {*}  {Promise<any>}
   */
  const getApiSpec = async (): Promise<any> => {
    const { data } = await api.get('/api/api-docs/');

    return data;
  };

  return {
    getProjects,
    getProject,
    getTemplate,
    createProject,
    createActivity,
    getAllCodes,
    getApiSpec
  };
};
