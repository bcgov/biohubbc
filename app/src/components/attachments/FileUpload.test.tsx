import { fireEvent, render, waitFor } from '@testing-library/react';
import { APIError } from 'hooks/api/useAxios';
import { useBiohubApi } from 'hooks/useBioHubApi';
import React from 'react';
import FileUpload from './FileUpload';

jest.mock('../../hooks/useBioHubApi');
const mockUseBiohubApi = {
  project: {
    uploadProjectArtifacts: jest.fn<Promise<any>, []>()
  }
};

const mockBiohubApi = ((useBiohubApi as unknown) as jest.Mock<typeof mockUseBiohubApi>).mockReturnValue(
  mockUseBiohubApi
);

const projectId = 1;

const renderContainer = () => {
  return render(<FileUpload projectId={projectId} />);
};

describe('FileUpload', () => {
  it('matches the snapshot', () => {
    const { asFragment } = renderContainer();

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles file upload success', async () => {
    let resolveRef: (value: unknown) => void;

    const mockUploadPromise = new Promise(function (resolve: any, reject: any) {
      resolveRef = resolve;
    });

    mockBiohubApi().project.uploadProjectArtifacts.mockReturnValue(mockUploadPromise);

    const { asFragment, getByTestId, getByText } = renderContainer();

    const testFile = new File(['test png content'], 'testpng.txt', { type: 'text/plain' });

    const dropZoneInput = getByTestId('drop-zone-input');

    fireEvent.change(dropZoneInput, { target: { files: [testFile] } });

    await waitFor(() => {
      expect(mockBiohubApi().project.uploadProjectArtifacts).toHaveBeenCalledWith(
        projectId,
        [testFile],
        expect.any(Object),
        expect.any(Function)
      );

      expect(getByText('testpng.txt')).toBeVisible();

      expect(getByText('uploading')).toBeVisible();
    });

    // Manually trigger the upload resolve to simulate a successful upload
    // @ts-ignore
    resolveRef(null);

    await waitFor(() => {
      expect(getByText('complete')).toBeVisible();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles file upload failure', async () => {
    let rejectRef: (reason: unknown) => void;

    const mockUploadPromise = new Promise(function (resolve: any, reject: any) {
      rejectRef = reject;
    });

    mockBiohubApi().project.uploadProjectArtifacts.mockReturnValue(mockUploadPromise);

    const { asFragment, getByTestId, getByText } = renderContainer();

    const testFile = new File(['test png content'], 'testpng.txt', { type: 'text/plain' });

    const dropZoneInput = getByTestId('drop-zone-input');

    fireEvent.change(dropZoneInput, { target: { files: [testFile] } });

    await waitFor(() => {
      expect(mockBiohubApi().project.uploadProjectArtifacts).toHaveBeenCalledWith(
        projectId,
        [testFile],
        expect.any(Object),
        expect.any(Function)
      );

      expect(getByText('testpng.txt')).toBeVisible();

      expect(getByText('uploading')).toBeVisible();
    });

    // Manually trigger the upload reject to simulate an unsuccessful upload
    // @ts-ignore
    rejectRef(new APIError({ response: { data: { message: 'File was evil!' } } } as any));

    await waitFor(() => {
      expect(getByText('File was evil!')).toBeVisible();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
