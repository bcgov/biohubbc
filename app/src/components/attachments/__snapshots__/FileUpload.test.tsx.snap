// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload handles file upload failure 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-37"
  >
    <div
      class="MuiBox-root MuiBox-root-38 makeStyles-dropZone-34"
    >
      <section>
        <div
          class="MuiBox-root MuiBox-root-40"
          tabindex="0"
        >
          <input
            autocomplete="off"
            data-testid="drop-zone-input"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <div
            class="MuiBox-root MuiBox-root-41"
          >
            <svg
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 12H4V17H20V12H22V17A2 2 0 0 1 20 19H4A2 2 0 0 1 2 17M11 5H13V8H16V10H13V13H11V10H8V8H11Z"
                style="fill: currentColor;"
              />
            </svg>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Drag your files here, or 
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-browseLink-39 MuiTypography-colorPrimary"
              >
                Browse Files
              </a>
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Maximum file size: 50 MB
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Maximum file count: 10
            </p>
          </div>
        </div>
      </section>
    </div>
    <div
      class="MuiBox-root MuiBox-root-42"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-root makeStyles-uploadListItem-35 MuiListItem-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root-53"
          >
            <div
              class="MuiBox-root MuiBox-root-54"
            >
              <div
                class="MuiBox-root MuiBox-root-55"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  testpng.txt
                </p>
                <p
                  class="MuiTypography-root makeStyles-uploadListStatus-36 MuiTypography-body1"
                >
                  File was evil!
                </p>
              </div>
              <div
                class="MuiBox-root MuiBox-root-56"
              >
                <div
                  aria-valuemax="100"
                  aria-valuemin="0"
                  aria-valuenow="0"
                  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate"
                  role="progressbar"
                >
                  <div
                    class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
                    style="transform: translateX(-100%);"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-57"
            >
              <button
                aria-label="delete attachment"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                tabindex="0"
                title="Delete Attachment"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    role="presentation"
                    style="width: 1.5rem; height: 1.5rem;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M13.46,12L19,17.54V19H17.54L12,13.46L6.46,19H5V17.54L10.54,12L5,6.46V5H6.46L12,10.54L17.54,5H19V6.46L13.46,12Z"
                      style="fill: currentColor;"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FileUpload handles file upload success 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-13"
  >
    <div
      class="MuiBox-root MuiBox-root-14 makeStyles-dropZone-10"
    >
      <section>
        <div
          class="MuiBox-root MuiBox-root-16"
          tabindex="0"
        >
          <input
            autocomplete="off"
            data-testid="drop-zone-input"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <div
            class="MuiBox-root MuiBox-root-17"
          >
            <svg
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 12H4V17H20V12H22V17A2 2 0 0 1 20 19H4A2 2 0 0 1 2 17M11 5H13V8H16V10H13V13H11V10H8V8H11Z"
                style="fill: currentColor;"
              />
            </svg>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Drag your files here, or 
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-browseLink-15 MuiTypography-colorPrimary"
              >
                Browse Files
              </a>
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Maximum file size: 50 MB
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Maximum file count: 10
            </p>
          </div>
        </div>
      </section>
    </div>
    <div
      class="MuiBox-root MuiBox-root-18"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-root makeStyles-uploadListItem-11 MuiListItem-gutters"
        >
          <div
            class="MuiBox-root MuiBox-root-29"
          >
            <div
              class="MuiBox-root MuiBox-root-30"
            >
              <div
                class="MuiBox-root MuiBox-root-31"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  testpng.txt
                </p>
                <p
                  class="MuiTypography-root makeStyles-uploadListStatus-12 MuiTypography-body1"
                >
                  complete
                </p>
              </div>
              <div
                class="MuiBox-root MuiBox-root-32"
              >
                <div
                  aria-valuemax="100"
                  aria-valuemin="0"
                  aria-valuenow="100"
                  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate"
                  role="progressbar"
                >
                  <div
                    class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
                    style="transform: translateX(0%);"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-33"
            >
              <button
                aria-label="delete attachment"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                tabindex="0"
                title="Delete Attachment"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    role="presentation"
                    style="width: 1.5rem; height: 1.5rem;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M13.46,12L19,17.54V19H17.54L12,13.46L6.46,19H5V17.54L10.54,12L5,6.46V5H6.46L12,10.54L17.54,5H19V6.46L13.46,12Z"
                      style="fill: currentColor;"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FileUpload matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-4"
  >
    <div
      class="MuiBox-root MuiBox-root-5 makeStyles-dropZone-1"
    >
      <section>
        <div
          class="MuiBox-root MuiBox-root-7"
          tabindex="0"
        >
          <input
            autocomplete="off"
            data-testid="drop-zone-input"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          <div
            class="MuiBox-root MuiBox-root-8"
          >
            <svg
              role="presentation"
              style="width: 1.5rem; height: 1.5rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 12H4V17H20V12H22V17A2 2 0 0 1 20 19H4A2 2 0 0 1 2 17M11 5H13V8H16V10H13V13H11V10H8V8H11Z"
                style="fill: currentColor;"
              />
            </svg>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Drag your files here, or 
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-browseLink-6 MuiTypography-colorPrimary"
              >
                Browse Files
              </a>
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Maximum file size: 50 MB
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Maximum file count: 10
            </p>
          </div>
        </div>
      </section>
    </div>
    <div
      class="MuiBox-root MuiBox-root-9"
    />
  </div>
</DocumentFragment>
`;
